### PR TITLE
Add FE API /api/sql to support http query

### PR DESCRIPTION
**Pull request by mchades**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Solve #3978 

## Problem Summary(Required) ：

Add FE API /api/sql to support execute sql of SHOW, EXPLAIN, SELECT

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 61 / 192 (31.77%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/http/JsonSerializer.java|13|78|16.67%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|15|54|27.78%|
|:large_blue_circle:|com/starrocks/http/rest/ExecuteSqlAction.java|25|52|48.08%|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/qe/OriginStatement.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/http/HttpServer.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/http/rest/RestBaseAction.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 61 / 192 (31.77%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/http/JsonSerializer.java|13|78|16.67%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|15|54|27.78%|
|:large_blue_circle:|com/starrocks/http/rest/ExecuteSqlAction.java|25|52|48.08%|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/qe/OriginStatement.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/http/HttpServer.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/http/rest/RestBaseAction.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java [hunk 1]
```diff
@@ -712,54 +726,76 @@ private void handleQueryStmt(List<PlanFragment> fragments, List<ScanNode> scanNo
...
-        //
-        // 2. If this is a query, send the result expr fields first, and send result data back to client.
         RowBatch batch;
-        MysqlChannel channel = context.getMysqlChannel();
-        boolean isOutfileQuery = false;
+        boolean isOutfileQuery;
         if (queryStmt instanceof QueryStmt) {
             isOutfileQuery = ((QueryStmt) queryStmt).hasOutFileClause();
-        } else if (queryStmt instanceof QueryStatement) {
+        } else {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java:733]: The change in StmtExecutor seems not elegant.

File: fe/fe-core/src/main/java/com/starrocks/http/rest/ExecuteSqlAction.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        String originStmt = request.getContent();
+        if (Strings.isNullOrEmpty(originStmt)) {
+            response.appendContent(new RestBaseResult("query can not empty").toJson());
+            writeResponse(request, response, BAD_REQUEST);
+            return;
+        }
+
+        ConnectContext context = ConnectContext.get();
+        SqlScanner input = new SqlScanner(new StringReader(originStmt), context.getSessionVariable().getSqlMode());
+        com.starrocks.analysis.SqlParser parser = new SqlParser(input);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/http/rest/ExecuteSqlAction.java:62]: should use new parser.

File: fe/fe-core/src/main/java/com/starrocks/http/rest/ExecuteSqlAction.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    public ExecuteSqlAction(ActionController controller) {
+        super(controller);
+    }
+
+    public static void registerAction(ActionController controller) throws IllegalArgException {
+        controller.registerHandler(HttpMethod.POST, "/api/sql", new ExecuteSqlAction(controller));
+    }
+
+    @Override
+    protected void executeWithoutPassword(BaseRequest request, BaseResponse response) throws DdlException {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/http/rest/ExecuteSqlAction.java:51]: we should have the authentication and authentication

File: fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java (hunk 1 above)
Comment (mchades) [reply fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java:733]: Did you mean I should extract the process of outfileQuery judgment into individual functions?

File: fe/fe-core/src/main/java/com/starrocks/http/rest/ExecuteSqlAction.java (hunk 2 above)
Comment (mchades) [reply fe/fe-core/src/main/java/com/starrocks/http/rest/ExecuteSqlAction.java]: done

File: fe/fe-core/src/main/java/com/starrocks/http/rest/ExecuteSqlAction.java (hunk 3 above)
Comment (mchades) [reply fe/fe-core/src/main/java/com/starrocks/http/rest/ExecuteSqlAction.java:48]: Did you means privilege check? PrivilegeChecker added now.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 105 / 251 (41.83%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/http/JsonSerializer.java|28|118|23.73%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|24|57|42.11%|
|:large_blue_circle:|com/starrocks/http/rest/ExecuteSqlAction.java|45|68|66.18%|
|:large_blue_circle:|com/starrocks/qe/OriginStatement.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/http/rest/RestBaseAction.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/http/HttpServer.java|1|1|100.00%|

**Comment by xxhZs** [comment]
Hello, I would like to use this API to retrieve the schema of a table through HTTP. Where can I find an example for this?

**Pull request closed without merging**